rclpy/rclpy/parameter_event_handler.py

```diff
@@ -0,0 +1,208 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections import defaultdict
+from itertools import chain
+from multiprocessing import Lock
+from typing import Callable, Dict, Iterable, List, Optional, Tuple
+
+
+from rcl_interfaces.msg import ParameterEvent
+from rclpy.callback_groups import CallbackGroup
+from rclpy.event_handler import SubscriptionEventCallbacks
+from rclpy.node import Node
+from rclpy.parameter import Parameter
+from rclpy.qos import qos_profile_parameter_events
+from rclpy.qos import QoSProfile
+from rclpy.qos_overriding_options import QoSOverridingOptions
+
+
+class ParameterCallbackHandle:
+    def __init__(
+        self,
+        parameter_name: str,
+        node_name: str,
+        callback: Callable[[Parameter], None],
+    ):
+        self.parameter_name = parameter_name
+        self.node_name = node_name
+        self.callback = callback
+
+        self.mutex = Lock()
+
+
+class ParameterEventCallbackHandle:
+
+    def __init__(self, callback: Callable[[ParameterEvent], None]):
+        self.callback = callback
+        self.mutex = Lock()
+
+
+class ParameterEventHandler:
+    class Callbacks:
+
+        def __init__(self):
+            self.parameter_callbacks: Dict[
+                Tuple[str, str], List[ParameterCallbackHandle]
+            ] = defaultdict(list)
+            self.event_callbacks: List[ParameterEventCallbackHandle] = []
+            self.mutex = Lock()
+
+        def event_callback(self, event: ParameterEvent):
+            with self.mutex:
+                for (param_name, node_name), callbacks_list in self.parameter_callbacks.items():
+                    if parameter := ParameterEventHandler.get_parameter_from_event(
+                        event, parameter_name=param_name, node_name=node_name
+                    ):
+                        for callback_handle in reversed(callbacks_list):
+                            with callback_handle.mutex:
+                                callback_handle.callback(parameter)
+
+                for event_callback in reversed(self.event_callbacks):
+                    with event_callback.mutex:
+                        event_callback.callback(event)
+
+        def add_parameter_callback(
+            self,
+            parameter_name: str,
+            node_name: str,
+            callback: Callable[[Parameter], None],
+        ) -> ParameterCallbackHandle:
+            handle = ParameterCallbackHandle(
```

Comment:
Here I think we are missing the equivalent of https://github.com/ros2/rclcpp/blob/6e97990a32afa17ab98ef4c800783c5cce801786/rclcpp/src/rclcpp/parameter_event_handler.cpp#L65 .  That's needed because the node name coming from the callback will be fully qualified, so we need to do the same here.

Reply:
This comment is still open.

Reply:
This is still open as well.

Reply:
Oh, I totally missed that.  Thank you for the explanation.